tuning/sft_trainer.py

```diff
@@ -29,6 +31,40 @@ def on_save(self, args, state, control, **kwargs):
             os.remove(os.path.join(checkpoint_path, "pytorch_model.bin"))
 
 
+class FileLoggingCallback(TrainerCallback):
+    """Exports metrics, e.g., training loss to a file in the checkpoint directory."""
+    # Keys to be exported from the log dict
+    log_keys = ["loss", "epoch"]
+
+    def __init__(self, logger):
+        self.logger = logger
+        self.existing_logs = []
+
+    def on_log(self, args, state, control, logs=None, **kwargs):
+        """Checks if this log contains keys of interest, e.g., los, and if so, creates
+        train_loss.json in the model output dir (if it doesn't already exist),
+        appends the subdict of the log & dumps the file.
+        """
+
+        log_file_path = os.path.join(args.output_dir, "train_loss.json")
+        if logs is not None:
+            try:
+                # Take the subdict of the last log line; if any log_keys aren't part of this log
+                # object, asssume this line is something else, e.g., train completion, and skip.
+                log_obj = {k: logs[k] for k in FileLoggingCallback.log_keys}
+            except KeyError:
+                return
+
+            # Redump the json file in the checkpoint directory with the updated log line
+            self.existing_logs.append(log_obj)
+            with open(log_file_path, "w") as log_file:
+                json.dump(self.existing_logs, log_file, sort_keys=True, indent=4)
```

Comment:
One thing I would be a bit worried about is the need to open and close file repeatedly. Also looks like we are overriding the existing logs instead of appending, any particular reason for that?

Reply:
Just that this is writing a JSON file, so we need to reparse it to add to the list. We could alternatively write a list of JSON objects (i.e., each line is a JSON object representing one log), and in that case just append per log?

I agree that it's a bit of a bummer to open and close the file so much, although it's still probably quite small compared to the actual training of the model, and the parsing is the main problem. If we write a json object per log, we can keep it open in append mode and flush on each written log I guess?